File: activemq-core/src/main/java/org/activemq/openwire/OpenWireFormat.java
/**
* <a href="http://activemq.org">ActiveMQ: The Open Source Message Fabric</a>
*
* Copyright 2005 (C) LogicBlaze, Inc. http://www.logicblaze.com
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*
**/
package org.activemq.openwire;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.lang.reflect.Method;
import java.util.HashMap;

import org.activeio.ByteArrayOutputStream;
import org.activeio.ByteSequence;
import org.activeio.Packet;
import org.activeio.adapter.PacketToInputStream;
import org.activeio.command.ClassLoading;
import org.activeio.command.WireFormat;
import org.activeio.packet.ByteArrayPacket;
import org.activemq.command.CommandTypes;
import org.activemq.command.DataStructure;
import org.activemq.command.MarshallAware;

/**
 * 
 * @version $Revision$
 */
final public class OpenWireFormat implements WireFormat {
    
    static final byte NULL_TYPE = CommandTypes.NULL;
    private static final int MARSHAL_CACHE_SIZE = Short.MAX_VALUE/2;
    private DataStreamMarshaller dataMarshallers[];
    private int version;
    private boolean stackTraceEnabled=true;
    private boolean tcpNoDelayEnabled=false;
    private boolean cacheEnabled=true;

    private HashMap marshallCacheMap = new HashMap();
    private short nextMarshallCacheIndex=0;    
    private short lasMarshallCacheEvictionIndex=100;    
    private DataStructure marshallCache[] = new DataStructure[MARSHAL_CACHE_SIZE];
    private DataStructure unmarshallCache[] = new DataStructure[MARSHAL_CACHE_SIZE];
    
    public OpenWireFormat() {
        this(true);
    }
    
    public OpenWireFormat(boolean cacheEnabled) {
        setVersion(1);
        setCacheEnabled(cacheEnabled);
    }
    
    public int hashCode() {
        return  version 
            ^ (cacheEnabled?0x10000000:0x20000000)
            ^ (stackTraceEnabled?0x30000000:0x40000000);
    }
    
    public boolean equals(Object object) {
        if( object == null )
            return false;
        OpenWireFormat o = (OpenWireFormat) object;
        return o.stackTraceEnabled == stackTraceEnabled &&
            o.cacheEnabled == cacheEnabled &&
            o.version == version;
    }
    
    public String toString() {
        return "OpenWireFormat{version="+version+", cacheEnabled="+cacheEnabled+", stackTraceEnabled="+stackTraceEnabled+"}";
    }
    
    public int getVersion() {
        return version;
    }
    
    public Packet marshal(Object command) throws IOException {
        
        MarshallAware ma=null;
        // If not using value caching, then the marshaled form is always the same
        if( !cacheEnabled && ((DataStructure)command).isMarshallAware() ) {
            ma = (MarshallAware) command;
        }
        
        ByteSequence sequence=null;
        if( ma!=null ) {
            sequence = ma.getCachedMarshalledForm(this);
        }
        
        if( sequence == null ) {
            
            int size=1;
            if( command != null) {
                
                DataStructure c = (DataStructure) command;
                byte type = c.getDataStructureType();
                DataStreamMarshaller dsm = (DataStreamMarshaller) dataMarshallers[type & 0xFF];
                if( dsm == null )
                    throw new IOException("Unknown data type: "+type);
                BooleanStream bs = new BooleanStream();
                size += dsm.marshal1(this, c, bs);
                size += bs.marshalledSize();

                ByteArrayOutputStream baos = new ByteArrayOutputStream(size);
                DataOutputStream ds = new DataOutputStream(baos);
                ds.writeInt(size);
                ds.writeByte(type);
                bs.marshal(ds);
                dsm.marshal2(this, c, ds, bs);                
                ds.close();
                
                sequence = baos.toByteSequence();
                
            } else {
                
                ByteArrayOutputStream baos = new ByteArrayOutputStream(5);
                DataOutputStream daos = new DataOutputStream(baos);
                daos.writeInt(size);
                daos.writeByte(NULL_TYPE);
                daos.close();
                sequence = baos.toByteSequence();
            }
            
            if( ma!=null ) {
                ma.setCachedMarshalledForm(this, sequence);
            }
        }
        return new ByteArrayPacket(sequence);
    }
    
    public Object unmarshal(Packet packet) throws IOException {
        ByteSequence sequence = packet.asByteSequence();
        DataInputStream dis = new DataInputStream(new PacketToInputStream(packet));
        int size = dis.readInt();
        if( sequence.getLength() != size+4 )
            System.out.println("Packet size does not match marshaled size: "+size+", "+(sequence.getLength()-4));
//            throw new IOException("Packet size does not match marshaled size");        
        Object command = doUnmarshal(dis);
        if( !cacheEnabled && ((DataStructure)command).isMarshallAware() ) {
            ((MarshallAware) command).setCachedMarshalledForm(this, sequence);
        }
        return command;
    }
    
    public void marshal(Object o, DataOutputStream ds) throws IOException {
        int size=1;
        if( o != null) {
            DataStructure c = (DataStructure) o;
            byte type = c.getDataStructureType();
            DataStreamMarshaller dsm = (DataStreamMarshaller) dataMarshallers[type & 0xFF];
            if( dsm == null )
                throw new IOException("Unknown data type: "+type);

            BooleanStream bs = new BooleanStream();
            size += dsm.marshal1(this, c, bs);
            size += bs.marshalledSize(); 

            ds.writeInt(size);
            ds.writeByte(type);            
            bs.marshal(ds);
            dsm.marshal2(this, c, ds, bs);
        } else {
            ds.writeInt(size);
            ds.writeByte(NULL_TYPE);
        }
    }
    
    public Object unmarshal(DataInputStream dis) throws IOException {
        dis.readInt();
        return doUnmarshal(dis);
    }
    
    /**
     * Allows you to dynamically switch the version of the openwire protocol being used.
     * @param version
     */
    public void setVersion(int version) {
        String mfName = "org.activemq.openwire.v"+version+".MarshallerFactory";
        Class mfClass;
        try {
            mfClass = ClassLoading.loadClass(mfName, getClass().getClassLoader());
        } catch (ClassNotFoundException e) {
            throw (IllegalArgumentException)new IllegalArgumentException("Invalid version: "+version+", could not load "+mfName).initCause(e);
        }
        try {
            Method method = mfClass.getMethod("createMarshallerMap", new Class[]{OpenWireFormat.class});
            dataMarshallers = (DataStreamMarshaller[]) method.invoke(null, new Object[]{this});
        } catch (Throwable e) {
            throw (IllegalArgumentException)new IllegalArgumentException("Invalid version: "+version+", "+mfName+" does not properly implement the createMarshallerMap method.").initCause(e);
        }
        this.version = version;
    }
        
    public Object doUnmarshal(DataInputStream dis) throws IOException {
        byte dataType = dis.readByte();
        if( dataType!=NULL_TYPE ) {
            DataStreamMarshaller dsm = (DataStreamMarshaller) dataMarshallers[dataType & 0xFF];
            if( dsm == null )
                throw new IOException("Unknown data type: "+dataType);
            Object data = dsm.createObject();
            BooleanStream bs = new BooleanStream();
            bs.unmarshal(dis);
            dsm.unmarshal(this, data, dis, bs);
            return data;
        } else {
            return null;
        }
    }
    
    public int marshal1NestedObject(DataStructure o, BooleanStream bs) throws IOException {
        bs.writeBoolean(o != null);
        if( o == null ) 
            return 0;

        if( o.isMarshallAware() ) {
            MarshallAware ma = (MarshallAware) o;
            ByteSequence sequence=ma.getCachedMarshalledForm(this);
            bs.writeBoolean(sequence!=null);
            if( sequence!=null ) {
                return 1 + sequence.getLength();           
            }
        }
        
        byte type = o.getDataStructureType();
        DataStreamMarshaller dsm = (DataStreamMarshaller) dataMarshallers[type & 0xFF];
        if( dsm == null )
            throw new IOException("Unknown data type: "+type);
        return 1 + dsm.marshal1(this, o, bs);
    }
    
    public void marshal2NestedObject(DataStructure o, DataOutputStream ds, BooleanStream bs) throws IOException {
        if( !bs.readBoolean() ) 
            return;
            
        byte type = o.getDataStructureType();
        ds.writeByte(type);

        if( o.isMarshallAware() && bs.readBoolean() ) {
                        
            MarshallAware ma = (MarshallAware) o;
            ByteSequence sequence=ma.getCachedMarshalledForm(this);
            ds.write(sequence.getData(), sequence.getOffset(), sequence.getLength());
            
        } else {
            
            DataStreamMarshaller dsm = (DataStreamMarshaller) dataMarshallers[type & 0xFF];
            if( dsm == null )
                throw new IOException("Unknown data type: "+type);
            dsm.marshal2(this, o, ds, bs);
            
        }
    }
    
    public DataStructure unmarshalNestedObject(DataInputStream dis, BooleanStream bs) throws IOException {
        if( bs.readBoolean() ) {
            
            byte dataType = dis.readByte();
            DataStreamMarshaller dsm = (DataStreamMarshaller) dataMarshallers[dataType & 0xFF];
            if( dsm == null )
                throw new IOException("Unknown data type: "+dataType);
            DataStructure data = dsm.createObject();

            if( data.isMarshallAware() && bs.readBoolean() ) {
                
                dis.readInt();
                dis.readByte();
                
                BooleanStream bs2 = new BooleanStream();
                bs2.unmarshal(dis);
                dsm.unmarshal(this, data, dis, bs2);

                // TODO: extract the sequence from the dis and associate it.
//                MarshallAware ma = (MarshallAware)data
//                ma.setCachedMarshalledForm(this, sequence);
                
            } else {
                dsm.unmarshal(this, data, dis, bs);
            }
            
            return data;
        } else {
            return null;
        }
    }
    
    public Short getMarshallCacheIndex(Object o) {
        return (Short) marshallCacheMap.get(o);
    }
    
    public Short addToMarshallCache(Object o) {
        Short index = new Short(nextMarshallCacheIndex++);
        if( nextMarshallCacheIndex >= MARSHAL_CACHE_SIZE ) {
            nextMarshallCacheIndex=0;
        }
        lasMarshallCacheEvictionIndex++;
        if( lasMarshallCacheEvictionIndex >= MARSHAL_CACHE_SIZE ) {
            lasMarshallCacheEvictionIndex=0;
        }
        if( marshallCache[lasMarshallCacheEvictionIndex]!=null ) {
            marshallCacheMap.remove(marshallCache[lasMarshallCacheEvictionIndex]);
            marshallCache[lasMarshallCacheEvictionIndex]=null;
        }
        marshallCacheMap.put(o, index);
        return index;
    }
    
    public void setInUnmarshallCache(short index, DataStructure o) {
        unmarshallCache[index]=o;
    }
    
    public DataStructure getFromUnmarshallCache(short index) {
        return unmarshallCache[index];
    }


    public void setStackTraceEnabled(boolean b) {
        stackTraceEnabled = b;
    }
    public boolean isStackTraceEnabled() {
        return stackTraceEnabled;
    }

    public boolean isTcpNoDelayEnabled() {
        return tcpNoDelayEnabled;
    }
    public void setTcpNoDelayEnabled(boolean tcpNoDelayEnabled) {
        this.tcpNoDelayEnabled = tcpNoDelayEnabled;
    }

    public boolean isCacheEnabled() {
        return cacheEnabled;
    }
    public void setCacheEnabled(boolean cacheEnabled) {
        this.cacheEnabled = cacheEnabled;
    }

}

File: openwire-dotnet/src/OpenWire.Client/Core/OpenWireFormat.cs
/*
 * Copyright 2006 The Apache Software Foundation or its licensors, as
 * applicable.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System;
using System.IO;

using OpenWire.Client.Commands;
using OpenWire.Client.Core;
using OpenWire.Client.IO;

namespace OpenWire.Client.Core
{
    /// <summary>
    /// Represents the wire format
    /// </summary>
    public class OpenWireFormat
    {
        static private char[] MAGIC = new char[] { 'A', 'c', 't', 'i', 'v', 'e', 'M', 'Q' };
        
        private BaseDataStreamMarshaller[] dataMarshallers;
        private const byte NULL_TYPE = 0;
        private WireFormatInfo wireFormatInfo = new WireFormatInfo();
        
        public OpenWireFormat()
        {
            // lets configure the wire format
            wireFormatInfo.Magic = CreateMagicBytes();
            wireFormatInfo.Version = 1;
            wireFormatInfo.StackTraceEnabled = true;
            wireFormatInfo.TcpNoDelayEnabled = true;
			wireFormatInfo.PrefixPacketSize = true;
			wireFormatInfo.TightEncodingEnabled = true;
            
            dataMarshallers = new BaseDataStreamMarshaller[256];
            MarshallerFactory factory = new MarshallerFactory();
            factory.configure(this);
        }
        
        public WireFormatInfo WireFormatInfo {
            get {
                return wireFormatInfo;
            }
        }
        
        public bool StackTraceEnabled {
            get {
                return wireFormatInfo.StackTraceEnabled;
            }
        }
        
        public void addMarshaller(BaseDataStreamMarshaller marshaller)
        {
            byte type = marshaller.GetDataStructureType();
            dataMarshallers[type & 0xFF] = marshaller;
        }
        
        public void Marshal(Object o, BinaryWriter ds)
        {
            int size = 1;
            if (o != null)
            {
                DataStructure c = (DataStructure) o;
                byte type = c.GetDataStructureType();
                BaseDataStreamMarshaller dsm = (BaseDataStreamMarshaller) dataMarshallers[type & 0xFF];
                if (dsm == null)
                    throw new IOException("Unknown data type: " + type);
                
                BooleanStream bs = new BooleanStream();
                size += dsm.TightMarshal1(this, c, bs);
                size += bs.MarshalledSize();
                
                BaseDataStreamMarshaller.WriteInt(size, ds);
                BaseDataStreamMarshaller.WriteByte(type, ds);
                bs.Marshal(ds);
                dsm.TightMarshal2(this, c, ds, bs);
            }
            else
            {
                BaseDataStreamMarshaller.WriteInt(size, ds);
                BaseDataStreamMarshaller.WriteByte(NULL_TYPE, ds);
            }
        }
        
        public Object Unmarshal(BinaryReader dis)
        {
            // lets ignore the size of the packet
            BaseDataStreamMarshaller.ReadInt(dis);
            
            // first byte is the type of the packet
            byte dataType = BaseDataStreamMarshaller.ReadByte(dis);
            if (dataType != NULL_TYPE)
            {
                BaseDataStreamMarshaller dsm = (BaseDataStreamMarshaller) dataMarshallers[dataType & 0xFF];
                if (dsm == null)
                    throw new IOException("Unknown data type: " + dataType);
                //Console.WriteLine("Parsing type: " + dataType + " with: " + dsm);
                Object data = dsm.CreateObject();
                BooleanStream bs = new BooleanStream();
                bs.Unmarshal(dis);
                dsm.TightUnmarshal(this, data, dis, bs);
                return data;
            }
            else
            {
                return null;
            }
        }
        
        public int TightMarshalNestedObject1(DataStructure o, BooleanStream bs)
        {
            bs.WriteBoolean(o != null);
            if (o == null)
                return 0;
            
            if (o.IsMarshallAware())
            {
                MarshallAware ma = (MarshallAware) o;
                byte[] sequence = ma.GetMarshalledForm(this);
                bs.WriteBoolean(sequence != null);
                if (sequence != null)
                {
                    return 1 + sequence.Length;
                }
            }
            
            byte type = o.GetDataStructureType();
            if (type == 0) {
                throw new IOException("No valid data structure type for: " + o + " of type: " + o.GetType());
            }
            BaseDataStreamMarshaller dsm = (BaseDataStreamMarshaller) dataMarshallers[type & 0xFF];
            if (dsm == null)
                throw new IOException("Unknown data type: " + type);
            //Console.WriteLine("Marshalling type: " + type + " with structure: " + o);
            return 1 + dsm.TightMarshal1(this, o, bs);
        }
        
        public void TightMarshalNestedObject2(DataStructure o, BinaryWriter ds, BooleanStream bs)
        {
            if (!bs.ReadBoolean())
                return ;
            
            byte type = o.GetDataStructureType();
            BaseDataStreamMarshaller.WriteByte(type, ds);
            
            if (o.IsMarshallAware() && bs.ReadBoolean())
            {
                MarshallAware ma = (MarshallAware) o;
                byte[] sequence = ma.GetMarshalledForm(this);
                ds.Write(sequence, 0, sequence.Length);
            }
            else
            {
                
                BaseDataStreamMarshaller dsm = (BaseDataStreamMarshaller) dataMarshallers[type & 0xFF];
                if (dsm == null)
                    throw new IOException("Unknown data type: " + type);
                dsm.TightMarshal2(this, o, ds, bs);
            }
        }
        
        public DataStructure TightUnmarshalNestedObject(BinaryReader dis, BooleanStream bs)
        {
            if (bs.ReadBoolean())
            {
                
                byte dataType = BaseDataStreamMarshaller.ReadByte(dis);
                BaseDataStreamMarshaller dsm = (BaseDataStreamMarshaller) dataMarshallers[dataType & 0xFF];
                if (dsm == null)
                    throw new IOException("Unknown data type: " + dataType);
                DataStructure data = dsm.CreateObject();
                
                if (data.IsMarshallAware() && bs.ReadBoolean())
                {
                    BaseDataStreamMarshaller.ReadInt(dis);
                    BaseDataStreamMarshaller.ReadByte(dis);
                    
                    BooleanStream bs2 = new BooleanStream();
                    bs2.Unmarshal(dis);
                    dsm.TightUnmarshal(this, data, dis, bs2);
                    
                    // TODO: extract the sequence from the dis and associate it.
                    //                MarshallAware ma = (MarshallAware)data
                    //                ma.setCachedMarshalledForm(this, sequence);
                }
                else
                {
                    dsm.TightUnmarshal(this, data, dis, bs);
                }
                
                return data;
            }
            else
            {
                return null;
            }
        }
        
        /// <summary>
        /// Method CreateMagicBytes
        /// </summary>
        private byte[] CreateMagicBytes()
        {
            byte[] answer = new byte[MAGIC.Length];
            for (int i = 0; i < answer.Length; i++)
            {
                answer[i] = (byte) MAGIC[i];
            }
            return answer;
        }
    }
}
